### PR TITLE
Migrate from libsuperuser to libsu

### LINKED ISSUE #3033: /data/data folder fails to list all subfolders from v3.6.6
**Describe the bug**
From version 3.6.6, `/data/data` folder only lists few app data folders (with root access). No error message or anything, it simply fails to lists all appdata folders inside `/data/data`.  It was working properly until version 3.6.5

**To Reproduce**
1. Go to '/data/data/' folder

**Expected behavior**
Should lists all subfolders since it has root access

**Smartphone (please complete the following information):**
 - Device: Redmi K30
 - OS: 11
 - Rooted: Yes
 - Version: 3.6.7

### PR DESCRIPTION
**Pull request by TranceLove**
## Description
Fixes #3033.

Original [libsuperuser](https://github.com/Chainfire/libsuperuser) had stopped development since 2019, and even chainfire suggested migration to newer libs like [libsu](https://github.com/topjohnwu/libsu), by topjohnwu the author of Magisk.

- Removed `MainActivity.shellInteractive` - libsu's Shell object is already a Singleton
- Removed `MainActivity`'s HandlerThread for su shell - libsu's Shell itself will handle the threading
- Removed ability to run callback for Shell commands. libsu's callback feature is truly async, that our current codebase won't handle nicely

Also, added `mount-master` flag as mentioned by https://github.com/topjohnwu/libsu/issues/75, to fix problems that only few directories are displayed when listing `/data/data` directory.

~~This is still a draft - a couple of TODO:~~
- ~~use libsu I/O classes for accessing root files, for TextEditorActivity and DatabaseViewerActivity continue to work~~

#### Automatic tests
- [ ] Added test cases
  
#### Manual tests
- [x] Done  
  
- Device: Pixel 5 emulator, running Android 12
- Device: Fairphone 3, running LineageOS 16.0 (9.0)

Root prompt, list root directories should continue working. Also listing `/data/data` should be same as `ls -la /data/data` from rooted adb shell.

#### Build tasks success  

Successfully running following tasks on local:
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

### CODE REVIEW

**Comment by VishalNehra** [comment]
@TranceLove Much awaited 🙌

File: app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java [hunk 1]
```diff
@@ -1387,12 +1378,12 @@ protected void onDestroy() {
...
-      if (SDK_INT >= JELLY_BEAN_MR2) {
-        // let it finish up first with what it's doing
-        handlerThread.quitSafely();
-      } else handlerThread.quit();
-      shellInteractive.close();
+      // close interactive shell
+      try {
+        Shell.getShell().close();
+      } catch (IOException e) {
+        Log.w(TAG, "Error closing Shell", e);
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java:1385]: This is not a warning, its an error.

**Comment by EmmanuelMess** [comment]
Testing on my rooted Moto RAZR D3:
Create file ".txt" in data, try to edit it with text editor and get a crash.

```
2022-02-11 14:04:37.523 1800-1800/com.amaze.filemanager.debug W/MMUMapper: fail to register MVA, unsupported format(0x5)
2022-02-11 14:04:37.529 1800-1800/com.amaze.filemanager.debug E/ReadTextFileTask: Error on text read
    java.io.FileNotFoundException: /storage/sdcard0/Android/data/com.amaze.filemanager.debug/cache/.txt: open failed: ENOENT (No such file or directory)
        at libcore.io.IoBridge.open(IoBridge.java:451)
        at java.io.FileInputStream.<init>(FileInputStream.java:78)
        at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.loadFile(ReadTextFileCallable.java:136)
        at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:86)
        at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:46)
        at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:39)
        at io.reactivex.Flowable.subscribe(Flowable.java:14826)
        at io.reactivex.Flowable.subscribe(Flowable.java:14773)
        at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
        at java.util.concurrent.FutureTask.run(FutureTask.java:137)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
        at java.lang.Thread.run(Thread.java:856)
     Caused by: libcore.io.ErrnoException: open failed: ENOENT (No such file or directory)
        at libcore.io.Posix.open(Native Method)
        at libcore.io.BlockGuardOs.open(BlockGuardOs.java:110)
        at libcore.io.IoBridge.open(IoBridge.java:435)
        at java.io.FileInputStream.<init>(FileInputStream.java:78) 
        at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.loadFile(ReadTextFileCallable.java:136) 
        at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:86) 
        at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:46) 
        at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:39) 
        at io.reactivex.Flowable.subscribe(Flowable.java:14826) 
        at io.reactivex.Flowable.subscribe(Flowable.java:14773) 
        at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82) 
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66) 
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57) 
        at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:137) 
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150) 
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569) 
        at java.lang.Thread.run(Thread.java:856) 
```

**Comment by VishalNehra** [comment]
> Testing on my rooted Moto RAZR D3: Create file ".txt" in /data, try to edit it with text editor and get a crash.
> 
> ```
> 2022-02-11 14:04:37.523 1800-1800/com.amaze.filemanager.debug W/MMUMapper: fail to register MVA, unsupported format(0x5)
> 2022-02-11 14:04:37.529 1800-1800/com.amaze.filemanager.debug E/ReadTextFileTask: Error on text read
>     java.io.FileNotFoundException: /storage/sdcard0/Android/data/com.amaze.filemanager.debug/cache/.txt: open failed: ENOENT (No such file or directory)
>         at libcore.io.IoBridge.open(IoBridge.java:451)
>         at java.io.FileInputStream.<init>(FileInputStream.java:78)
>         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.loadFile(ReadTextFileCallable.java:136)
>         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:86)
>         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:46)
>         at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:39)
>         at io.reactivex.Flowable.subscribe(Flowable.java:14826)
>         at io.reactivex.Flowable.subscribe(Flowable.java:14773)
>         at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
>         at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
>         at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
>         at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
>         at java.util.concurrent.FutureTask.run(FutureTask.java:137)
>         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
>         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
>         at java.lang.Thread.run(Thread.java:856)
>      Caused by: libcore.io.ErrnoException: open failed: ENOENT (No such file or directory)
>         at libcore.io.Posix.open(Native Method)
>         at libcore.io.BlockGuardOs.open(BlockGuardOs.java:110)
>         at libcore.io.IoBridge.open(IoBridge.java:435)
>         at java.io.FileInputStream.<init>(FileInputStream.java:78) 
>         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.loadFile(ReadTextFileCallable.java:136) 
>         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:86) 
>         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:46) 
>         at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:39) 
>         at io.reactivex.Flowable.subscribe(Flowable.java:14826) 
>         at io.reactivex.Flowable.subscribe(Flowable.java:14773) 
>         at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82) 
>         at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66) 
>         at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57) 
>         at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305) 
>         at java.util.concurrent.FutureTask.run(FutureTask.java:137) 
>         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150) 
>         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264) 
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076) 
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569) 
>         at java.lang.Thread.run(Thread.java:856) 
> ```
> 
> And try to delete the file:
> 
> ```
> 2022-02-11 14:13:14.425 1800-1800/com.amaze.filemanager.debug W/MMUMapper: fail to register MVA, unsupported format(0x5)
> 2022-02-11 14:13:14.433 1800-1800/com.amaze.filemanager.debug W/System.err: java.io.FileNotFoundException: /data/.DummyFile: open failed: EACCES (Permission denied)
> 2022-02-11 14:13:14.434 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:451)
> 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at java.io.FileOutputStream.<init>(FileOutputStream.java:88)
> 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.FileProperties.isWritable(FileProperties.kt:85)
> 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.checkFolder(MainActivityHelper.java:470)
> 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.deleteFiles(MainActivityHelper.java:669)
> 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.ui.dialogs.GeneralDialogCreation.lambda$deleteFilesDialog$1(GeneralDialogCreation.java:228)
> 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.ui.dialogs.GeneralDialogCreation$$ExternalSyntheticLambda4.onClick(Unknown Source)
> 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.afollestad.materialdialogs.MaterialDialog.onClick(MaterialDialog.java:410)
> 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.view.View.performClick(View.java:4091)
> 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.view.View$PerformClick.run(View.java:17081)
> 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.os.Handler.handleCallback(Handler.java:615)
> 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:92)
> 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.os.Looper.loop(Looper.java:153)
> 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:4987)
> 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at java.lang.reflect.Method.invokeNative(Native Method)
> 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at java.lang.reflect.Method.invoke(Method.java:511)
> 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:821)
> 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:584)
> 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err:     at dalvik.system.NativeStart.main(Native Method)
> 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err: Caused by: libcore.io.ErrnoException: open failed: EACCES (Permission denied)
> 2022-02-11 14:13:14.440 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.Posix.open(Native Method)
> 2022-02-11 14:13:14.440 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.BlockGuardOs.open(BlockGuardOs.java:110)
> 2022-02-11 14:13:14.441 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:435)
> 2022-02-11 14:13:14.441 1800-1800/com.amaze.filemanager.debug W/System.err: 	... 18 more
> 2022-02-11 14:13:14.461 1800-1800/com.amaze.filemanager.debug W/MMUMapper: invalid operation for unregister MVA with VA(0x57356000) size(354816) f(0x5)
> 2022-02-11 14:13:14.462 1800-1800/com.amaze.filemanager.debug W/MMUMapper: invalid operation for unregister MVA with VA(0x579bb000) size(354816) f(0x5)
> 2022-02-11 14:13:14.462 1800-1800/com.amaze.filemanager.debug W/MMUMapper: invalid operation for unregister MVA with VA(0x57dd7000) size(354816) f(0x5)
> ```

Doesn't seem to be a root directory.

**Comment by EmmanuelMess** [comment]
> > Testing on my rooted Moto RAZR D3: Create file ".txt" in /data, try to edit it with text editor and get a crash.
> > ```
> > 2022-02-11 14:04:37.523 1800-1800/com.amaze.filemanager.debug W/MMUMapper: fail to register MVA, unsupported format(0x5)
> > 2022-02-11 14:04:37.529 1800-1800/com.amaze.filemanager.debug E/ReadTextFileTask: Error on text read
> >     java.io.FileNotFoundException: /storage/sdcard0/Android/data/com.amaze.filemanager.debug/cache/.txt: open failed: ENOENT (No such file or directory)
> >         at libcore.io.IoBridge.open(IoBridge.java:451)
> >         at java.io.FileInputStream.<init>(FileInputStream.java:78)
> >         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.loadFile(ReadTextFileCallable.java:136)
> >         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:86)
> >         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:46)
> >         at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:39)
> >         at io.reactivex.Flowable.subscribe(Flowable.java:14826)
> >         at io.reactivex.Flowable.subscribe(Flowable.java:14773)
> >         at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
> >         at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
> >         at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
> >         at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
> >         at java.util.concurrent.FutureTask.run(FutureTask.java:137)
> >         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
> >         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
> >         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
> >         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
> >         at java.lang.Thread.run(Thread.java:856)
> >      Caused by: libcore.io.ErrnoException: open failed: ENOENT (No such file or directory)
> >         at libcore.io.Posix.open(Native Method)
> >         at libcore.io.BlockGuardOs.open(BlockGuardOs.java:110)
> >         at libcore.io.IoBridge.open(IoBridge.java:435)
> >         at java.io.FileInputStream.<init>(FileInputStream.java:78) 
> >         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.loadFile(ReadTextFileCallable.java:136) 
> >         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:86) 
> >         at com.amaze.filemanager.asynchronous.asynctasks.texteditor.read.ReadTextFileCallable.call(ReadTextFileCallable.java:46) 
> >         at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:39) 
> >         at io.reactivex.Flowable.subscribe(Flowable.java:14826) 
> >         at io.reactivex.Flowable.subscribe(Flowable.java:14773) 
> >         at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82) 
> >         at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66) 
> >         at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57) 
> >         at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305) 
> >         at java.util.concurrent.FutureTask.run(FutureTask.java:137) 
> >         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150) 
> >         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264) 
> >         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076) 
> >         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569) 
> >         at java.lang.Thread.run(Thread.java:856) 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And try to delete the file:
> > ```
> > 2022-02-11 14:13:14.425 1800-1800/com.amaze.filemanager.debug W/MMUMapper: fail to register MVA, unsupported format(0x5)
> > 2022-02-11 14:13:14.433 1800-1800/com.amaze.filemanager.debug W/System.err: java.io.FileNotFoundException: /data/.DummyFile: open failed: EACCES (Permission denied)
> > 2022-02-11 14:13:14.434 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:451)
> > 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at java.io.FileOutputStream.<init>(FileOutputStream.java:88)
> > 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.FileProperties.isWritable(FileProperties.kt:85)
> > 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.checkFolder(MainActivityHelper.java:470)
> > 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.deleteFiles(MainActivityHelper.java:669)
> > 2022-02-11 14:13:14.436 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.ui.dialogs.GeneralDialogCreation.lambda$deleteFilesDialog$1(GeneralDialogCreation.java:228)
> > 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.ui.dialogs.GeneralDialogCreation$$ExternalSyntheticLambda4.onClick(Unknown Source)
> > 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.afollestad.materialdialogs.MaterialDialog.onClick(MaterialDialog.java:410)
> > 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.view.View.performClick(View.java:4091)
> > 2022-02-11 14:13:14.437 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.view.View$PerformClick.run(View.java:17081)
> > 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.os.Handler.handleCallback(Handler.java:615)
> > 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:92)
> > 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.os.Looper.loop(Looper.java:153)
> > 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:4987)
> > 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at java.lang.reflect.Method.invokeNative(Native Method)
> > 2022-02-11 14:13:14.438 1800-1800/com.amaze.filemanager.debug W/System.err:     at java.lang.reflect.Method.invoke(Method.java:511)
> > 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:821)
> > 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:584)
> > 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err:     at dalvik.system.NativeStart.main(Native Method)
> > 2022-02-11 14:13:14.439 1800-1800/com.amaze.filemanager.debug W/System.err: Caused by: libcore.io.ErrnoException: open failed: EACCES (Permission denied)
> > 2022-02-11 14:13:14.440 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.Posix.open(Native Method)
> > 2022-02-11 14:13:14.440 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.BlockGuardOs.open(BlockGuardOs.java:110)
> > 2022-02-11 14:13:14.441 1800-1800/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:435)
> > 2022-02-11 14:13:14.441 1800-1800/com.amaze.filemanager.debug W/System.err: 	... 18 more
> > 2022-02-11 14:13:14.461 1800-1800/com.amaze.filemanager.debug W/MMUMapper: invalid operation for unregister MVA with VA(0x57356000) size(354816) f(0x5)
> > 2022-02-11 14:13:14.462 1800-1800/com.amaze.filemanager.debug W/MMUMapper: invalid operation for unregister MVA with VA(0x579bb000) size(354816) f(0x5)
> > 2022-02-11 14:13:14.462 1800-1800/com.amaze.filemanager.debug W/MMUMapper: invalid operation for unregister MVA with VA(0x57dd7000) size(354816) f(0x5)
> > ```
> 
> Doesn't seem to be a root directory.

What do you mean? The file was created in "/data", is /data not a root dir?

**Comment by VishalNehra** [comment]
java.io.FileNotFoundException: /storage/sdcard0/Android/data/com.amaze.filemanager.debug/cache/.txt: open failed: ENOENT (No such file or directory)
        at libcore.io.IoBridge.open(IoBridge.java:451)

The first crash path doesn't seem to be root directory

**Comment by TranceLove** [comment]
I was skipping using libsu's I/O package in favour of what we have been using. So it seems it becomes inevitable... :/

**Comment by TranceLove** [comment]
With ea98c2b I just add code to delete previous cached file before proceeding.

Writing back to root storage was not working on Pixel 5 emulator running Android 12. Not sure why yet, but YMMV IMO.

**Comment by EmmanuelMess** [comment]
> Writing back to root storage was not working on Pixel 5 emulator running Android 12. Not sure why yet, but YMMV IMO.

On my Moto it isn't working either, is this a known issue in develop?

**Review by EmmanuelMess — APPROVED** [review]

**Comment by TranceLove** [comment]
Writing back to root actually was **working** when running on Fairphone 3 with LineageOS 16.0 (9.0), as well as Oneplus 2 running LineageOS 18 (11).

Then back to stock rom, as I go with the `adb shell` path for debugging, `mount` actually didn't work as the prerequisite to write back to root - it gave an exit status of `1`.

```bash
airwave@hsv010 ~/android-sdk-linux/platform-tools $ ./adb -s emulator-5554 shell
emulator64_x86_64_arm64:/ $ su --mount-master
emulator64_x86_64_arm64:/ # mount -o remount,rw /system
mount: '/system' not in /proc/mounts
1|emulator64_x86_64_arm64:/ # mount -o remount,rw /                                                                                                                                                                                                  
'/dev/block/dm-4' is read-only
emulator64_x86_64_arm64:/ # 
```

Some references:

https://mobile.twitter.com/topjohnwu/status/1170404631865778177
https://forum.xda-developers.com/t/mount-system-as-rw-on-android-10-with-magisk-20-4.4081043/

So seems it's a no-go for stock roms...

And there is alternate way to make `mount -o remount, rw` possible... but that's entirely out of our app's scope.
https://forum.xda-developers.com/t/script-android-10-universal-mount-systemrw-superrw-featuring-makerw-read-write.4247311/

**Review by VishalNehra — APPROVED** [review]

**Pull request merged**